malariagen_data/ag3.py

```diff
@@ -2489,6 +2494,98 @@ def aa_allele_frequencies(
 
         return df_aaf
 
+    def plot_frequencies_heatmap(
+        self, df, index=None, max_len=100, y_label=None, colorbar=True, width=None
+    ):
+
+        """Plot a heatmap from a pandas DataFrame of frequencies, e.g., output from
+            `Ag3.snp_allele_frequencies()` or `Ag3.gene_cnv_frequencies()`. It's recommended to
+            filter the input DataFrame to just rows of interest, i.e., fewer rows than `max_len`.
+
+        Parameters
+        ----------
+        df : pandas DataFrame
+           A DataFrame of frequencies, e.g., output from `snp_allele_frequencies()` or `gene_cnv_frequencies()`.
+        index : str or list of str
+            One or more column headers that are present in the input dataframe. This becomes the heatmap y-axis
+            row labels. The column/s must produce a unique index.
+        max_len : int, optional
+            Displaying large styled dataframes may cause ipython notebooks to crash.
+        y_label : str, optional
+            This is the y-axis label that will be displayed on the heatmap.
+        colorbar : bool, optional
+            If False, colorbar is not output.
+        width : int, optional
+            Width of heatmap output.
+
+        """
+
+        # check len of input
+        if len(df) > max_len:
+            raise ValueError(f"Input DataFrame is longer than {max_len}")
+
+        if index is None:
+            index_names = list(df.index.names)
+            df = df.reset_index().copy()
+            index_col = (
+                df[index_names]
+                .astype(str)
+                .apply(lambda row: "_".join(row), axis="columns")
+            )
+
+        if index is not None:
+            df = df.reset_index().copy()
+            if isinstance(index, list):
+                index_col = (
+                    df[index]
+                    .astype(str)
+                    .apply(lambda row: "_".join(row), axis="columns")
+                )
+            if isinstance(index, str):
+                index_col = df[index].astype(str)
```

Comment:
```suggestion
        if index is None:
            index = list(df.index.names)

        df = df.reset_index().copy()
        if isinstance(index, list):
            index_col = (
                df[index]
                .astype(str)
                .apply(lambda row: "_".join(row), axis="columns")
            )
        elif isinstance(index, str):
            index_col = df[index].astype(str)
        else:
            raise TypeError("wrong type for index parameter, expected list or str")
```

Possible simplification.